src/shogun/mathematics/linalg/linalgVector.cpp

```diff
@@ -0,0 +1,168 @@
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/linalgVector.h>
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+template<class T>
+LinalgVector<T>::LinalgVector()
+{
+	init();
+}
+
+template<class T>
+LinalgVector<T>::LinalgVector(SGVector<T> const &vector)
+{
+	init();
+	m_data = reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen));
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data);
+	m_len = vector.vlen;
+}
+
+template<class T>
+LinalgVector<T>::LinalgVector(LinalgVector<T> const &vector)
+{
+	init();
+	m_data = vector.m_data;
+	m_len = vector.m_len;
+	m_onGPU = vector.m_onGPU;
+
+	if (vector.onGPU())
+	{
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+	}
+}
+
+template<class T>
+LinalgVector<T>::~LinalgVector()
+{
+	free(m_data);
+}
+
+template<class T>
+LinalgVector<T>& LinalgVector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen));
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data);
+	m_len = vector.vlen;
+	m_onGPU = false;
+	m_gpu_impl.release();
+	return *this;
+}
+
+template<class T>
+LinalgVector<T>& LinalgVector<T>::operator=(LinalgVector<T> const &vector)
+{
+
+	m_data = vector.m_data;
+	m_len = vector.m_len;
+	m_onGPU = vector.m_onGPU;
+	if (vector.onGPU())
+	{
+		m_gpu_impl.reset(new GPUVectorImpl(*(vector.m_gpu_impl)));
+	}
+	return *this;
+
+}
+
+template<class T>
+LinalgVector<T>::operator SGVector<T>() const
+{
+	return SGVector<T>(m_data, m_len);
```

Comment:
What about the case when data is on GPU? Say, you transfer the data on GPU once and then did a bunch of in-place operations. Then the data pointed by m_data is no-longer what the user would want. 

I think if the data is on GPU, it should copy that to m_data first. Then it should create a new SGVector, deep copy m_data to the vector of that SGVector and then return. This thing above will lead to segfault. For example:
```c++
SGVector<float64_t> v;
if (some_condition)
{
    // create a LinalgVector instance, vec, and do some linalg, finally assign that to the above SGVector
    v  = vec;
    // vec instance dies, along with its m_data
}
// try to use v, bam! dangling pointer, v.vector :(
```